flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java

```diff
@@ -163,8 +164,15 @@ YarnLocalResourceDescriptor registerSingleLocalResource(
         addToRemotePaths(whetherToAddToRemotePaths, resourcePath);
 
         if (Utils.isRemotePath(resourcePath.toString())) {
-            final FileStatus fileStatus = fileSystem.getFileStatus(resourcePath);
-            LOG.debug("Using remote file {} to register local resource", fileStatus.getPath());
+            final FileSystem srcFs = resourcePath.getFileSystem(fileSystem.getConf());
+            final FileStatus resourceFileStatus = srcFs.getFileStatus(resourcePath);
+            LOG.debug("Using remote file {} to register local resource", resourceFileStatus.getPath());
+            final Path applicationDir = getApplicationDirPath(homeDir, applicationId);
+            final String suffix = (relativeDstPath.isEmpty() ? "" : relativeDstPath + "/") + resourcePath.getName();
+            final Path dst = new Path(applicationDir, suffix);
+            LOG.debug("Copying from {} to {} with replication factor {}", resourcePath, dst, fileReplication);
+            FileUtil.copy(srcFs, resourcePath, fileSystem, dst, false, fileSystem.getConf());
```

Comment:
Won't this also copy files on the remote FS if they are already located on this FS?

Reply:
It is not necessary if the `srcPath` and `dstPath` have the sameschema, for example: form hdfs to hdfs. Otherwise, It needs to copy files between different FS. as follows:
![image](https://user-images.githubusercontent.com/13013780/103750230-73c02280-5041-11eb-8261-988afb631635.png)

Reply:
The schema alone might not be enough because the paths might refer to different hdfs filesystems. But maybe we can do this optimization as a follow up.